akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBProjectionImpl.scala

```diff
@@ -556,54 +564,69 @@ private[projection] object DynamoDBProjectionImpl {
                       .recoverWith { exc =>
                         logReplayException(logPrefix, originalEventEnvelope, fromSeqNr, exc)
                         Future.failed(exc)
                       }
                   case None =>
-                    Future.successful(
-                      triggerReplayIfPossible(
-                        sourceProvider,
-                        persistenceId,
-                        fromSeqNr,
-                        originalEventEnvelope.sequenceNr))
+                    if (triggerReplayIfPossible(
+                          sourceProvider,
+                          persistenceId,
+                          fromSeqNr,
+                          originalEventEnvelope.sequenceNr))
+                      FutureReplayTriggered
+                    else FutureNotReplayed
                 }
               }
 
             case _ =>
-              triggerReplayIfPossible(sourceProvider, offsetStore, originalEnvelope)
+              triggerReplayIfPossible(sourceProvider, offsetStore, originalEnvelope).map {
+                case true  => ReplayTriggered
+                case false => NotReplayed
+              }(ExecutionContext.parasitic)
           }
         case _ =>
-          FutureFalse // no replay support for non typed envelopes
+          FutureNotReplayed // no replay support for non typed envelopes
       }
     }
 
     def observer(context: ProjectionContext): HandlerObserver[Envelope] =
       context.asInstanceOf[ProjectionContextImpl[Offset, Envelope]].observer
 
+    def accepted(envelope: Envelope, context: ProjectionContext): Future[Option[(Envelope, ProjectionContext)]] = {
+      if (isFilteredEvent(envelope) && settings.warnAboutFilteredEventsInFlow) {
+        log.info("atLeastOnceFlow doesn't support skipping envelopes. Envelope [{}] still emitted.", envelope)
+      }
+      loadEnvelope(envelope, sourceProvider).map { loadedEnvelope =>
+        offsetStore.addInflight(loadedEnvelope)
+        Some((loadedEnvelope, context))
+      }
+    }
+
     val validate = Flow[(Envelope, ProjectionContext)]
       .mapAsync(1) {
         case (env, context) =>
           offsetStore
             .validate(env)
             .flatMap {
               case Accepted =>
-                if (isFilteredEvent(env) && settings.warnAboutFilteredEventsInFlow) {
-                  log.info("atLeastOnceFlow doesn't support skipping envelopes. Envelope [{}] still emitted.", env)
-                }
-                loadEnvelope(env, sourceProvider).map { loadedEnvelope =>
-                  offsetStore.addInflight(loadedEnvelope)
-                  Some((loadedEnvelope, context))
-                }
+                accepted(env, context)
               case Duplicate =>
                 Future.successful(None)
               case RejectedSeqNr =>
-                replayIfPossible(env, observer(context)).map(_ => None)(ExecutionContext.parasitic)
```

Comment:
ah, here is the None, which means that it is filtered out and the offset will not be stored

Reply:
Yeah, and that was the right behaviour for triggered replays, but not for the immediate replays.